File: test/easy/find_lucky_test.dart
import 'package:test/test.dart';
import 'package:dark_coding_challenges/easy/find_lucky.dart';

class FindLuckyTestCases {
  List<int> arr;
  int expected;
  FindLuckyTestCases(this.arr, this.expected);
}

void main() {
  group("Feature:", () {
    final testCases = [
      FindLuckyTestCases([2, 2, 3, 4], 2),
      FindLuckyTestCases([1, 2, 2, 3, 3, 3], 3),
      FindLuckyTestCases([2, 2, 2, 3, 3], -1),
    ];

    for (final step in testCases) {
      final arr = step.arr;
      final expected = step.expected;
      test("\n\tGiven: the array $arr.\n\tThen: the maximum frequency relationship between the value and index is $expected.", () {
        int actual = findLucky(arr);
        expect(actual, equals(expected));
      });
    }
  });
}

File: test/easy/FindLuckyTest.php
<?php

declare(strict_types=1);

use Src\Easy\FindLucky;

dataset('findLuckyData', [
    [[2, 2, 3, 4], 2],
    [[1, 2, 2, 3, 3, 3], 3],
    [[2, 2, 2, 3, 3], -1],
    [[0, 0, 0], -1],
    [[1], 1],
    [[2], -1],
    [[2, 2, 3, 3, 3], 3],
    [[1, 1, 2, 2], 2],
    [[4, 4, 4, 4], 4],
    [[7, 7, 7], -1],
    [[0, 1, 2, 2], 2],
]);
it('\nt\tGiven: the array :arr.\n\tThen: the maximum frequency of the relationship between the number and its index is :expected.', function (array $arr, int $expected) {
    $instance = new FindLucky;

    $result = $instance->solve($arr);

    expect($result)->toBe($expected);

})->with('findLuckyData');
